src/Operation/Equals.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+/**
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace loophp\collection\Operation;
+
+use Closure;
+use Generator;
+use Iterator;
+
+/**
+ * @immutable
+ *
+ * @template TKey
+ * @template T
+ */
+final class Equals extends AbstractOperation
+{
+    /**
+     * @pure
+     *
+     * @return Closure(Iterator<TKey, T>): Closure(Iterator<TKey, T>): Generator<int, bool>
+     */
+    public function __invoke(): Closure
+    {
+        return
+            /**
+             * @param Iterator<TKey, T> $other
+             *
+             * @return Closure(Iterator<TKey, T>): Generator<int, bool>
+             */
+            static function (Iterator $other): Closure {
+                /** @var Closure(Iterator<TKey, T>): Generator<int, bool> $pipe */
+                $pipe = Pipe::of()(
+                    Diff::of()(...$other),
+                    IsEmpty::of(),
+                );
+
+                // Point free style.
+                return $pipe;
```

Comment:
Be careful! [It's a trap!](https://i.kym-cdn.com/entries/icons/original/000/000/157/itsatrap.jpg) !

Using `Diff` means that you're going to go through the whole iterator.

But yet what you're trying to do is to compare if they are equal, in other words: find if there is **at least** two elements that are different.

It means that as soon as two elements are different, there is no need to go through the whole iterator, we know that they are different.

It's then a job for the `MatchOne` operation.

Reply:
Hey @drupol, thanks for this; does this work if the elements are in different positions in the collections?

As in:

Collection (1, 2, 3)
Collection (2, 3, 1)

These should still be considered equal.

Reply:
Or are you thinking we should consider them unequal?

I feel like there's scope to have 2 operations maybe: `equals` and `same`

Reply:
> Hey @drupol, thanks for this; does this work if the elements are in different positions in the collections?
> 
> As in:
> 
> Collection (1, 2, 3)
> Collection (2, 3, 1)

No it doesn't do that sadly.



> I feel like there's scope to have 2 operations maybe: `equals` and `same`

Exact, we should have 2 different operations for that.

Reply:
Cool, in that case:
- `equals` would need to check the entire collection, so this implementation should be fine
- `same` should stop as soon as an element is different or the end of one collection is reached, because we want them in the same positions

Regarding allowing the user to customise the comparison behaviour:
I want at least one of the operations (I'm thinking `equals`) to _only accept a single parameter_, otherwise it won't work with PHPUnit's `assertObjectEquals` method: https://github.com/sebastianbergmann/phpunit/blob/master/src/Framework/Constraint/Object/ObjectEquals.php#L97-L102

Perhaps for `same` we can allow that with a signature: `Collection::same(Collection $other, ?callable $comparator = nul): bool`.

WDYT?

Reply:
It seems ok to me, I guess we will refine all of these during the review.